sw/device/silicon_creator/mask_rom/mask_rom_epmp_test.c

```diff
@@ -0,0 +1,392 @@
+  //
+  // Using a `call` here (`jal` or `jalr`) sets the return address (`ra`)
+  // register. When an exception is raised the interrupt handler will recover
+  // by restarting execution at the address in `ra` thereby making it appear
+  // as if the call returned normally.
+  //
+  //   ...
+  //   jal ra, pc # <- Set return address and jump to pc.
+  //   ...        # <- Interrupt handler restarts execution at the next
+  //              #    instruction in the caller, here.
+  //
+  // pc:
+  //   unimp      # <- Illegal instruction or access fault. Enter interrupt
+  //                   handler.
+  //
+  ((void (*)(void))pc)();
+
+  // Be careful to ensure that the exception was raised when trying to
+  // execute `pc` just in case a valid instruction is actually executed
+  // and then execution continued to a point where an exception is
+  // raised.
+  if (exception_received != kExceptionNone && exception_pc != (uintptr_t)pc) {
+    return false;
+  }
+  return exception_received == expect;
+}
+
+/**
+ * An instruction that has no bits set.
+ *
+ * Attempts to execute this instruction, `unimp`, will result in an illegal
+ * instruction exception.
+ *
+ * Note that if compressed instructions are enabled only the first two bytes
+ * will be decoded (as `c.unimp`).
+ */
+static const uint32_t kUnimpInstruction = 0;
+
+/**
+ * Illegal instruction residing in .rodata.
+ */
+static const uint32_t illegal_ins_ro[] = {
+    kUnimpInstruction,
+};
+
+/**
+ * Illegal instruction residing in .data.
+ */
+static uint32_t illegal_ins_rw[] = {
+    kUnimpInstruction,
+};
+
+/**
+ * Report whether the given pointer points to a location with the provided
+ * address space.
+ *
+ * @param ptr Pointer to test.
+ * @param start Address of the start of the address space.
+ * @param size The size of the address space in bytes.
+ * @returns Whether the pointer is in the address space.
+ */
+static bool is_in_address_space(const void *ptr, uintptr_t start,
+                                uintptr_t size) {
+  return (uintptr_t)ptr >= start && (uintptr_t)ptr < (start + size);
+}
+
+/**
+ * Enable execution of SRAM for the given controller.
+ *
+ * @param ctrl_addr Base address for SRAM controller.
+ * @returns Whether execution was enabled successfully or not.
+ */
+static bool sram_exec_enable(uint32_t ctrl_addr) {
+  // TODO: use the SRAM driver or DIF when available.
+  const uint32_t kEnableExecution = 5;
+  abs_mmio_write32(ctrl_addr + SRAM_CTRL_EXEC_REG_OFFSET, kEnableExecution);
+  return abs_mmio_read32(ctrl_addr + SRAM_CTRL_EXEC_REG_OFFSET) ==
+         kEnableExecution;
+}
+
+/**
+ * Set to false if a test fails.
+ */
+static bool passed = true;
+
+/**
+ * Custom CHECK macro to assert a condition that if false should cause the
+ * test to fail. Note: we can't use the normal CHECK macro because it tries to
+ * write to the DV address space but that is locked by the ePMP configuration.
+ */
+#define CHECK(condition)                  \
```

Comment:
probably something for later: can we use `RETURN_IF_ERROR` and `EXECUTE_TEST` in this file?